lib/src/Pos/Logic/Types.hs

```diff
@@ -27,53 +25,39 @@ import           Pos.Util.Chrono (NE, NewestFirst, OldestFirst)
 -- | The interface to a logic layer, i.e. some component which encapsulates
 -- blockchain / crypto logic.
 data Logic m = Logic
-    { -- The stakeholder id of our node.
+    { -- | The stakeholder id of our node.
       ourStakeholderId   :: StakeholderId
-      -- Get a block, perhaps from a database.
+      -- | Get a block, perhaps from a database.
     , getBlock           :: HeaderHash -> m (Maybe Block)
-      -- Stream blocks from first hash to second hash.
-      -- Conduit is chosen mainly due to precedent: it's already used in
-      -- cardano-sl.
-    , getChainFrom       :: HeaderHash
-                         -> Source m Block
-      -- Get a block header.
-      -- TBD: necessary? Is it any different/faster than getting the block
-      -- and taking the header?
+      -- | Get a block header.
     , getBlockHeader     :: HeaderHash -> m (Maybe BlockHeader)
-      -- Inspired by 'getHeadersFromManyTo'.
-      -- Included here because that function is quite complicated; it's not
-      -- clear whether it can be expressed simply in terms of getBlockHeader.:q
+      -- TODO CSL-2089 use conduits in this and the following methods
+      -- | Retrieve block header hashes from specified interval.
+    , getHashesRange     :: Maybe Word -- ^ Optional limit on how many to bring in.
+                         -> HeaderHash
+                         -> HeaderHash
+                         -> m (OldestFirst NE HeaderHash)
+      -- | Interface for 'getHeadersFromManyTo'. Retrieves blocks from
+      -- the checkpoints to some particular point (or tip, if
+      -- 'Nothing').
     , getBlockHeaders    :: Maybe Word -- ^ Optional limit on how many to bring in.
                          -> NonEmpty HeaderHash
                          -> Maybe HeaderHash
-                         -> m (Either GetBlockHeadersError (NewestFirst NE BlockHeader))
-      -- Inspired by 'getHeadersFromToIncl', which is apparently distinct from
-      -- 'getHeadersFromManyTo' (getBlockHeaders without the tick above).
-      -- FIXME we must unify these.
-      -- May want to think about giving a streaming-IO interface (pipes, conduit
-      -- or similar).
-    , getBlockHeaders'   :: Maybe Word -- ^ Optional limit on how many to bring in.
-                         -> HeaderHash
-                         -> HeaderHash
-                         -> m (Either GetBlockHeadersError (OldestFirst NE HeaderHash))
-      -- Get the current tip of chain.
-      -- It's not in Maybe, as getBlock is, because really there should always
-      -- be a tip, whereas trying to get a block that isn't in the database is
-      -- normal.
+                         -> m (NewestFirst NE BlockHeader)
+      -- | Compute LCA with the main chain.
+    , getLcaMainChain    :: OldestFirst NE BlockHeader -> m (Maybe HeaderHash)
```

Comment:
Notice that this type expresses the normal failure case: there's no LCA, so you get `Nothing`.

Compare at `getHashesRange`, which completely denies the normal failure case of not knowing one of the endpoints.